swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -807,6 +807,16 @@ predicate readStep(Node node1, ContentSet c, Node node2) {
     c.isSingleton(any(Content::ArrayContent ac))
   )
   or
+  exists(ForEachStmt for |
+    node1.asExpr() = for.getSequence() and
+    node2.asPattern() = for.getPattern() and
+    (
+      c.isSingleton(any(Content::ArrayContent ac))
+      or
+      c.isSingleton(any(Content::CollectionContent cc))
+    )
```

Comment:
I'm not super comfortable with this read step unconditionally reading both an `ArrayContent` _and_ a `CollectionContent`. This does make it seem like they should be unified. Alternatively, could we say that we read `ArrayContent` when the sequence is an array type, and otherwise we read a `CollectionContent`?

Reply:
I'm leaning more and more towards the view that we should unify `ArrayContent` and `CollectionContent` - not in this PR, but soon.  And make `ArrayElement` a deprecated alternative MAD syntax for `CollectionElement`.

Reply:
I agree. I think we should ask the other language teams why _they_ have two versions first, though.

Reply:
> Alternatively, could we say that we read `ArrayConten`t when the sequence is an array type, and otherwise we read a `CollectionContent`?

I think that's the best answer for now, and I'll reach out about why the other teams don't unify them.

Reply:
I'm actually increasingly keen to unify them.  There are all sorts of cases where we don't know whether something is an array or not, also having the separation seems to make subscript reads unreliable, and then there are classes like `ContiguousArray` which I'm not sure whether to consider as an "Array" or not.

Please `@mention` me into the cross-language discussion wherever it's going on as I'd like to understand if there are any good reasons not to do this.